File: backend/src/modules/mtree/mtree.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like } from 'typeorm';
import { Mtree } from './entities/mtree.entity';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';
import { KamailioRpcService } from '../../common/services/kamailio-rpc.service';

@Injectable()
export class MtreeService {
  constructor(
    @InjectRepository(Mtree)
    private mtreeRepository: Repository<Mtree>,
    private kamailioRpcService: KamailioRpcService,
  ) {}

  async findAll(dto: PaginationDto & { tname?: string }): Promise<PaginatedResult<Mtree>> {
    const { page, limit, keyword, tname } = dto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) where.tprefix = Like(`%${keyword}%`);
    if (tname) where.tname = tname;
    const [items, total] = await this.mtreeRepository.findAndCount({
      where, skip, take: limit, order: { tname: 'ASC', tprefix: 'ASC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async create(data: Partial<Mtree>): Promise<Mtree> {
    return this.mtreeRepository.save(this.mtreeRepository.create(data));
  }

  async update(id: number, data: Partial<Mtree>): Promise<Mtree> {
    const item = await this.mtreeRepository.findOne({ where: { id } });
    if (!item) throw new NotFoundException('记录不存在');
    Object.assign(item, data);
    return this.mtreeRepository.save(item);
  }

  async remove(id: number): Promise<void> {
    await this.mtreeRepository.delete(id);
  }

  async reload(tname: string): Promise<void> {
    await this.kamailioRpcService.call('mtree.reload', [tname]);
  }

  async match(tname: string, prefix: string): Promise<any> {
    return this.kamailioRpcService.call('mtree.match', [tname, prefix]);
  }

  async summary(tname?: string): Promise<any> {
    if (tname) {
      return this.kamailioRpcService.call('mtree.summary', [tname]);
    }
    return this.kamailioRpcService.call('mtree.summary');
  }
}

File: kam-admin-module-routing/src/main/java/io/kamailio/admin/modules/mtree/service/MtreeService.java
package io.kamailio.admin.modules.mtree.service;

import io.kamailio.admin.common.RpcTimeouts;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.common.service.KamailioRpcService;
import io.kamailio.admin.modules.mtree.entity.Mtree;
import io.kamailio.admin.modules.mtree.repository.MtreeRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class MtreeService {
    private final MtreeRepository repository;
    private final KamailioRpcService kamailioRpc;

    public MtreeService(MtreeRepository repository, KamailioRpcService kamailioRpc) {
        this.repository = repository;
        this.kamailioRpc = kamailioRpc;
    }

    public PaginatedResult<Mtree> findAll(PaginationDto dto, String tname) {
        var sort = Sort.by("tname", "tprefix");
        var pageable = PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), sort);
        Page<Mtree> page;
        if (tname != null && !tname.isBlank()) {
            page = repository.findByTname(tname, pageable);
        } else if (dto.getKeyword() != null && !dto.getKeyword().isBlank()) {
            page = repository.findByTprefixContaining(dto.getKeyword(), pageable);
        } else {
            page = repository.findAll(pageable);
        }
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    @Transactional
    public Mtree create(Mtree data) { return repository.save(data); }

    @Transactional
    public Mtree update(Integer id, Mtree data) {
        var item = repository.findById(id).orElseThrow(() -> new RuntimeException("记录不存在"));
        if (data.getTname() != null) item.setTname(data.getTname());
        if (data.getTprefix() != null) item.setTprefix(data.getTprefix());
        if (data.getTvalue() != null) item.setTvalue(data.getTvalue());
        return repository.save(item);
    }

    @Transactional
    public void remove(Integer id) { repository.deleteById(id); }

    public void reload(String tname) { kamailioRpc.mtreeReload(tname).block(RpcTimeouts.DEFAULT_BLOCK); }
    public Object match(String tname, String prefix) { return kamailioRpc.mtreeMatch(tname, prefix).block(RpcTimeouts.DEFAULT_BLOCK); }
    public Object summary(String tname) { return kamailioRpc.mtreeSummary(tname).block(RpcTimeouts.DEFAULT_BLOCK); }
}
